File: sdk/dart/lib/src/state.dart
/// Persisted client state and the throttling rules around it
/// (SPEC.md sections 12.2, 12.4, 12.7, and 16.2).
library;

import 'dart:convert';
import 'dart:io';

/// Whether an index with this sequence may be adopted (SPEC.md section 12.4).
///
/// Equal is accepted: re-fetching the same index is the ordinary case, not an
/// attack. Strictly smaller is refused, because a valid signature proves only
/// that the publisher issued the document at some point, never that it is the
/// newest one, so a network attacker could otherwise replay an old index to
/// steer clients onto a version with known holes.
///
/// A refusal here is not an error to report. Mirrors legitimately lag behind
/// one another, and surfacing that to the user as a failure would train them to
/// ignore a message that usually means nothing.
bool acceptsSequence(int sequence, int? lastSeenSequence) =>
    lastSeenSequence == null || sequence >= lastSeenSequence;

/// Stats for one candidate source key (URL or `service:<id>`).
class SourceStat {
  SourceStat({
    this.successes = 0,
    this.failures = 0,
    this.consecutiveFailures = 0,
    this.lastBytesPerSecond,
    this.lastSuccessAt,
    this.lastFailureAt,
  });

  factory SourceStat.fromJson(Map<String, dynamic> json) => SourceStat(
        successes: json['successes'] as int? ?? 0,
        failures: json['failures'] as int? ?? 0,
        consecutiveFailures: json['consecutiveFailures'] as int? ?? 0,
        lastBytesPerSecond: json['lastBytesPerSecond'] as int?,
        lastSuccessAt: switch (json['lastSuccessAt']) {
          final String value => DateTime.tryParse(value),
          _ => null,
        },
        lastFailureAt: switch (json['lastFailureAt']) {
          final String value => DateTime.tryParse(value),
          _ => null,
        },
      );

  int successes;
  int failures;
  int consecutiveFailures;
  int? lastBytesPerSecond;
  DateTime? lastSuccessAt;
  DateTime? lastFailureAt;

  Map<String, dynamic> toJson() => {
        'successes': successes,
        'failures': failures,
        'consecutiveFailures': consecutiveFailures,
        if (lastBytesPerSecond != null)
          'lastBytesPerSecond': lastBytesPerSecond,
        if (lastSuccessAt != null)
          'lastSuccessAt': lastSuccessAt!.toUtc().toIso8601String(),
        if (lastFailureAt != null)
          'lastFailureAt': lastFailureAt!.toUtc().toIso8601String(),
      };
}

/// What the client remembers between runs, per (product, channel).
class UpdateState {
  UpdateState({
    this.lastCheckAt,
    this.lastResult,
    this.lastSeenSequence,
    this.lastSeenFallbackSequence,
    this.lastSeenDirectorySequence,
    this.lastSuccessfulSourceKey,
    Map<String, SourceStat>? sourceStats,
    Set<int>? skipped,
  })  : sourceStats = sourceStats ?? <String, SourceStat>{},
        skipped = skipped ?? <int>{};

  factory UpdateState.fromJson(Map<String, dynamic> json) {
    final rawSkipped = json['skipped'];
    final rawStats = json['sourceStats'];
    final stats = <String, SourceStat>{};
    if (rawStats is Map) {
      for (final entry in rawStats.entries) {
        final value = entry.value;
        if (entry.key is String && value is Map<String, dynamic>) {
          stats[entry.key as String] = SourceStat.fromJson(value);
        } else if (entry.key is String && value is Map) {
          stats[entry.key as String] =
              SourceStat.fromJson(Map<String, dynamic>.from(value));
        }
      }
    }
    return UpdateState(
      lastCheckAt: switch (json['lastCheckAt']) {
        final String value => DateTime.tryParse(value),
        _ => null,
      },
      lastResult: json['lastResult'] as String?,
      lastSeenSequence: json['lastSeenSequence'] as int?,
      lastSeenFallbackSequence: json['lastSeenFallbackSequence'] as int?,
      lastSeenDirectorySequence: json['lastSeenDirectorySequence'] as int?,
      lastSuccessfulSourceKey: json['lastSuccessfulSourceKey'] as String?,
      sourceStats: stats,
      skipped: rawSkipped is List ? rawSkipped.whereType<int>().toSet() : null,
    );
  }

  DateTime? lastCheckAt;
  String? lastResult;

  /// The highest index sequence ever accepted. Never lower it.
  int? lastSeenSequence;

  /// The highest fallback sequence ever accepted (product-scoped anti-rollback).
  int? lastSeenFallbackSequence;

  /// The highest directory sequence ever accepted (product-scoped anti-rollback).
  int? lastSeenDirectorySequence;

  /// Key of the last candidate that fully succeeded (URL or `service:<id>`).
  String? lastSuccessfulSourceKey;

  /// Per-source outcomes from real attempts only (SPEC §12.7).
  final Map<String, SourceStat> sourceStats;

  /// Codes the user chose to skip. Ignored when an update is mandatory.
  final Set<int> skipped;

  Map<String, dynamic> toJson() => {
        if (lastCheckAt != null)
          'lastCheckAt': lastCheckAt!.toUtc().toIso8601String(),
        if (lastResult != null) 'lastResult': lastResult,
        if (lastSeenSequence != null) 'lastSeenSequence': lastSeenSequence,
        if (lastSeenFallbackSequence != null)
          'lastSeenFallbackSequence': lastSeenFallbackSequence,
        if (lastSeenDirectorySequence != null)
          'lastSeenDirectorySequence': lastSeenDirectorySequence,
        if (lastSuccessfulSourceKey != null)
          'lastSuccessfulSourceKey': lastSuccessfulSourceKey,
        if (sourceStats.isNotEmpty)
          'sourceStats': {
            for (final entry in sourceStats.entries)
              entry.key: entry.value.toJson(),
          },
        'skipped': skipped.toList()..sort(),
      };

  /// Records a newly accepted sequence, never moving the high-water mark down.
  void observeSequence(int sequence) {
    final seen = lastSeenSequence;
    if (seen == null || sequence > seen) lastSeenSequence = sequence;
  }

  void observeFallbackSequence(int sequence) {
    final seen = lastSeenFallbackSequence;
    if (seen == null || sequence > seen) lastSeenFallbackSequence = sequence;
  }

  void observeDirectorySequence(int sequence) {
    final seen = lastSeenDirectorySequence;
    if (seen == null || sequence > seen) lastSeenDirectorySequence = sequence;
  }

  void recordSourceSuccess(String key, {int? bytesPerSecond}) {
    if (key.isEmpty) return;
    final stat = sourceStats.putIfAbsent(key, SourceStat.new);
    stat
      ..successes += 1
      ..consecutiveFailures = 0
      ..lastSuccessAt = DateTime.now().toUtc();
    if (bytesPerSecond != null && bytesPerSecond > 0) {
      stat.lastBytesPerSecond = bytesPerSecond;
    }
    lastSuccessfulSourceKey = key;
  }

  void recordSourceFailure(String key) {
    if (key.isEmpty) return;
    final stat = sourceStats.putIfAbsent(key, SourceStat.new);
    stat
      ..failures += 1
      ..consecutiveFailures += 1
      ..lastFailureAt = DateTime.now().toUtc();
  }
}

abstract class UpdateStateStore {
  Future<UpdateState> load();

  Future<void> save(UpdateState state);
}

/// Keeps state in a JSON file, one per (product, channel).
///
/// Separate files rather than one shared document: two channels of the same app
/// have independent sequences, and mixing them would let a beta index hold back
/// a stable one.
class FileUpdateStateStore implements UpdateStateStore {
  FileUpdateStateStore({
    required Directory directory,
    required String product,
    required String channel,
  }) : file = File('${directory.path}${Platform.pathSeparator}'
            'rup-state-${_sanitize(product)}-${_sanitize(channel)}.json');

  final File file;

  static String _sanitize(String value) =>
      value.replaceAll(RegExp(r'[^A-Za-z0-9._-]'), '_');

  @override
  Future<UpdateState> load() async {
    try {
      if (!await file.exists()) return UpdateState();
      final decoded = json.decode(await file.readAsString());
      if (decoded is! Map<String, dynamic>) return UpdateState();
      return UpdateState.fromJson(decoded);
    } catch (_) {
      // Corrupt state must not block updating. The only thing lost is the
      // rollback high-water mark, which the next successful check restores;
      // refusing to run would be the worse failure, since it disables updates
      // permanently until someone deletes a file they do not know about.
      return UpdateState();
    }
  }

  @override
  Future<void> save(UpdateState state) async {
    await file.parent.create(recursive: true);
    final temp = File('${file.path}.tmp');
    await temp.writeAsString(json.encode(state.toJson()), flush: true);
    await temp.rename(file.path);
  }
}

/// In-memory store, for tests and for hosts that do not want a file.
class MemoryUpdateStateStore implements UpdateStateStore {
  MemoryUpdateStateStore([UpdateState? initial])
      : _state = initial ?? UpdateState();

  UpdateState _state;

  @override
  Future<UpdateState> load() async => _state;

  @override
  Future<void> save(UpdateState state) async => _state = state;
}

/// The intervals from SPEC.md section 12.2. Hosts may override them.
class UpdatePolicy {
  const UpdatePolicy({
    this.afterSuccess = const Duration(hours: 24),
    this.afterFailure = const Duration(hours: 1),
    this.documentTimeout = const Duration(seconds: 10),
    this.downloadIdleTimeout = const Duration(seconds: 60),
    this.downloadRetries = 3,
    this.downloadRetryBackoff = const Duration(milliseconds: 500),
    this.downloadConcurrency = 8,
    this.downloadChunkSize = 4 * 1024 * 1024,
  });

  final Duration afterSuccess;
  final Duration afterFailure;

  /// Applies to index and manifest fetches, which are small.
  final Duration documentTimeout;

  /// Applies between chunks of an artifact download, never to the whole
  /// transfer: a large package on a slow link legitimately takes minutes, so
  /// any total timeout is either useless or actively harmful.
  final Duration downloadIdleTimeout;

  /// Attempts per mirror URL for transient failures (network / 5xx / 408 / 429).
  final int downloadRetries;

  /// Initial backoff between retries; doubles after each failed attempt.
  final Duration downloadRetryBackoff;

  /// Parallel Range workers for one URL. `1` forces single-connection mode.
  final int downloadConcurrency;

  /// Preferred Range slice size in bytes (last slice is truncated to EOF).
  final int downloadChunkSize;

  /// Whether enough time has passed since the last check.
  bool shouldCheck(UpdateState state, {DateTime? now}) {
    final last = state.lastCheckAt;
    if (last == null) return true;
    final wait = state.lastResult == 'error' ? afterFailure : afterSuccess;
    return (now ?? DateTime.now()).difference(last) >= wait;
  }
}

File: sdk/node/src/state.ts
/**
 * Persisted client state and the throttling rules around it (SPEC.md sections
 * 12.2, 12.4, 12.7, and 16.2).
 */

import { mkdir, readFile, rename, writeFile } from "node:fs/promises";
import { dirname, join } from "node:path";

/**
 * Whether an index with this sequence may be adopted (SPEC.md section 12.4).
 *
 * Equal is accepted: re-fetching the same index is the ordinary case, not an
 * attack. Strictly smaller is refused, because a valid signature proves only
 * that the publisher issued the document at some point, never that it is the
 * newest one, so a network attacker could otherwise replay an old index to
 * steer clients onto a version with known holes.
 *
 * A refusal here is not an error to report. Mirrors legitimately lag behind one
 * another, and surfacing that to the user as a failure would train them to
 * ignore a message that usually means nothing.
 */
export function acceptsSequence(
  sequence: number,
  lastSeenSequence: number | null | undefined,
): boolean {
  return (
    lastSeenSequence === null ||
    lastSeenSequence === undefined ||
    sequence >= lastSeenSequence
  );
}

/** Stats for one candidate source key (URL or `service:<id>`). */
export interface SourceStat {
  successes: number;
  failures: number;
  consecutiveFailures: number;
  lastBytesPerSecond?: number;
  lastSuccessAt?: string;
  lastFailureAt?: string;
}

function emptyStat(): SourceStat {
  return { successes: 0, failures: 0, consecutiveFailures: 0 };
}

/** What the client remembers between runs, per (product, channel). */
export class UpdateState {
  lastCheckAt: Date | null = null;
  lastResult: string | null = null;

  /** The highest index sequence ever accepted. Never lower it. */
  lastSeenSequence: number | null = null;

  /** The highest fallback sequence ever accepted (product-scoped). */
  lastSeenFallbackSequence: number | null = null;

  /** The highest directory sequence ever accepted (product-scoped). */
  lastSeenDirectorySequence: number | null = null;

  /** Key of the last candidate that fully succeeded (URL or `service:<id>`). */
  lastSuccessfulSourceKey: string | null = null;

  /** Per-source outcomes from real attempts only (SPEC section 12.7). */
  readonly sourceStats = new Map<string, SourceStat>();

  /** Codes the user chose to skip. Ignored when an update is mandatory. */
  readonly skipped = new Set<number>();

  static fromJson(raw: unknown): UpdateState {
    const state = new UpdateState();
    if (typeof raw !== "object" || raw === null) return state;
    const json = raw as Record<string, unknown>;

    if (typeof json.lastCheckAt === "string") {
      const parsed = new Date(json.lastCheckAt);
      state.lastCheckAt = Number.isNaN(parsed.getTime()) ? null : parsed;
    }
    if (typeof json.lastResult === "string") state.lastResult = json.lastResult;
    if (typeof json.lastSeenSequence === "number") {
      state.lastSeenSequence = json.lastSeenSequence;
    }
    if (typeof json.lastSeenFallbackSequence === "number") {
      state.lastSeenFallbackSequence = json.lastSeenFallbackSequence;
    }
    if (typeof json.lastSeenDirectorySequence === "number") {
      state.lastSeenDirectorySequence = json.lastSeenDirectorySequence;
    }
    if (typeof json.lastSuccessfulSourceKey === "string") {
      state.lastSuccessfulSourceKey = json.lastSuccessfulSourceKey;
    }
    if (typeof json.sourceStats === "object" && json.sourceStats !== null) {
      for (const [key, value] of Object.entries(
        json.sourceStats as Record<string, unknown>,
      )) {
        if (typeof value !== "object" || value === null) continue;
        const entry = value as Record<string, unknown>;
        const stat = emptyStat();
        if (typeof entry.successes === "number") stat.successes = entry.successes;
        if (typeof entry.failures === "number") stat.failures = entry.failures;
        if (typeof entry.consecutiveFailures === "number") {
          stat.consecutiveFailures = entry.consecutiveFailures;
        }
        if (typeof entry.lastBytesPerSecond === "number") {
          stat.lastBytesPerSecond = entry.lastBytesPerSecond;
        }
        if (typeof entry.lastSuccessAt === "string") {
          stat.lastSuccessAt = entry.lastSuccessAt;
        }
        if (typeof entry.lastFailureAt === "string") {
          stat.lastFailureAt = entry.lastFailureAt;
        }
        state.sourceStats.set(key, stat);
      }
    }
    if (Array.isArray(json.skipped)) {
      for (const code of json.skipped) {
        if (typeof code === "number") state.skipped.add(code);
      }
    }
    return state;
  }

  toJson(): Record<string, unknown> {
    const out: Record<string, unknown> = {};
    if (this.lastCheckAt !== null) {
      out.lastCheckAt = this.lastCheckAt.toISOString();
    }
    if (this.lastResult !== null) out.lastResult = this.lastResult;
    if (this.lastSeenSequence !== null) {
      out.lastSeenSequence = this.lastSeenSequence;
    }
    if (this.lastSeenFallbackSequence !== null) {
      out.lastSeenFallbackSequence = this.lastSeenFallbackSequence;
    }
    if (this.lastSeenDirectorySequence !== null) {
      out.lastSeenDirectorySequence = this.lastSeenDirectorySequence;
    }
    if (this.lastSuccessfulSourceKey !== null) {
      out.lastSuccessfulSourceKey = this.lastSuccessfulSourceKey;
    }
    if (this.sourceStats.size > 0) {
      out.sourceStats = Object.fromEntries(this.sourceStats);
    }
    out.skipped = [...this.skipped].sort((a, b) => a - b);
    return out;
  }

  /** Records a newly accepted sequence, never moving the high-water mark down. */
  observeSequence(sequence: number): void {
    if (this.lastSeenSequence === null || sequence > this.lastSeenSequence) {
      this.lastSeenSequence = sequence;
    }
  }

  observeFallbackSequence(sequence: number): void {
    if (
      this.lastSeenFallbackSequence === null ||
      sequence > this.lastSeenFallbackSequence
    ) {
      this.lastSeenFallbackSequence = sequence;
    }
  }

  observeDirectorySequence(sequence: number): void {
    if (
      this.lastSeenDirectorySequence === null ||
      sequence > this.lastSeenDirectorySequence
    ) {
      this.lastSeenDirectorySequence = sequence;
    }
  }

  private statFor(key: string): SourceStat {
    let stat = this.sourceStats.get(key);
    if (stat === undefined) {
      stat = emptyStat();
      this.sourceStats.set(key, stat);
    }
    return stat;
  }

  recordSourceSuccess(key: string, bytesPerSecond?: number): void {
    if (key.length === 0) return;
    const stat = this.statFor(key);
    stat.successes += 1;
    stat.consecutiveFailures = 0;
    stat.lastSuccessAt = new Date().toISOString();
    if (bytesPerSecond !== undefined && bytesPerSecond > 0) {
      stat.lastBytesPerSecond = bytesPerSecond;
    }
    this.lastSuccessfulSourceKey = key;
  }

  recordSourceFailure(key: string): void {
    if (key.length === 0) return;
    const stat = this.statFor(key);
    stat.failures += 1;
    stat.consecutiveFailures += 1;
    stat.lastFailureAt = new Date().toISOString();
  }
}

export interface UpdateStateStore {
  load(): Promise<UpdateState>;
  save(state: UpdateState): Promise<void>;
}

const sanitize = (value: string): string =>
  value.replace(/[^A-Za-z0-9._-]/g, "_");

/**
 * Keeps state in a JSON file, one per (product, channel).
 *
 * Separate files rather than one shared document: two channels of the same app
 * have independent sequences, and mixing them would let a beta index hold back
 * a stable one.
 */
export class FileUpdateStateStore implements UpdateStateStore {
  readonly path: string;

  constructor(options: {
    directory: string;
    product: string;
    channel: string;
  }) {
    this.path = join(
      options.directory,
      `rup-state-${sanitize(options.product)}-${sanitize(options.channel)}.json`,
    );
  }

  async load(): Promise<UpdateState> {
    try {
      const raw = await readFile(this.path, "utf8");
      return UpdateState.fromJson(JSON.parse(raw));
    } catch {
      // Corrupt or missing state must not block updating. The only thing lost is
      // the rollback high-water mark, which the next successful check restores;
      // refusing to run would be the worse failure, since it disables updates
      // permanently until someone deletes a file they do not know about.
      return new UpdateState();
    }
  }

  async save(state: UpdateState): Promise<void> {
    await mkdir(dirname(this.path), { recursive: true });
    const temp = `${this.path}.tmp`;
    await writeFile(temp, JSON.stringify(state.toJson()), "utf8");
    await rename(temp, this.path);
  }
}

/** In-memory store, for tests and for hosts that do not want a file. */
export class MemoryUpdateStateStore implements UpdateStateStore {
  private state: UpdateState;

  constructor(initial?: UpdateState) {
    this.state = initial ?? new UpdateState();
  }

  async load(): Promise<UpdateState> {
    return this.state;
  }

  async save(state: UpdateState): Promise<void> {
    this.state = state;
  }
}

/** The intervals from SPEC.md section 12.2. Hosts may override them. */
export interface UpdatePolicy {
  /** Wait after a successful check. Default 24h. */
  afterSuccessMs: number;
  /** Wait after a failed check. Default 1h. */
  afterFailureMs: number;
  /** Applies to index and manifest fetches, which are small. Default 10s. */
  documentTimeoutMs: number;
  /**
   * Applies between chunks of an artifact download, never to the whole
   * transfer: a large package on a slow link legitimately takes minutes, so any
   * total timeout is either useless or actively harmful. Default 60s.
   */
  downloadIdleTimeoutMs: number;
  /** Attempts per mirror URL for transient failures. Default 3. */
  downloadRetries: number;
  /** Initial backoff between retries; doubles after each failure. Default 500ms. */
  downloadRetryBackoffMs: number;
  /** Parallel Range workers for one URL. `1` forces single-connection mode. */
  downloadConcurrency: number;
  /** Preferred Range slice size in bytes (last slice truncated to EOF). */
  downloadChunkSize: number;
}

export const defaultUpdatePolicy: UpdatePolicy = {
  afterSuccessMs: 24 * 60 * 60 * 1000,
  afterFailureMs: 60 * 60 * 1000,
  documentTimeoutMs: 10_000,
  downloadIdleTimeoutMs: 60_000,
  downloadRetries: 3,
  downloadRetryBackoffMs: 500,
  downloadConcurrency: 8,
  downloadChunkSize: 4 * 1024 * 1024,
};

export function resolvePolicy(policy?: Partial<UpdatePolicy>): UpdatePolicy {
  return { ...defaultUpdatePolicy, ...policy };
}

/** Whether enough time has passed since the last check. */
export function shouldCheck(
  policy: UpdatePolicy,
  state: UpdateState,
  now: Date = new Date(),
): boolean {
  const last = state.lastCheckAt;
  if (last === null) return true;
  const wait =
    state.lastResult === "error" ? policy.afterFailureMs : policy.afterSuccessMs;
  return now.getTime() - last.getTime() >= wait;
}
